### PR TITLE
Add auth support to the Reaper web UI [K8SSAND-1135]

### LINKED ISSUE #276: K8SSAND-1135 ⁃ Add support for authentication in Reaper web UI and REST API
This is a follow-up ticket after #170.

Reaper REST API and web UI are currently always unauthenticated. This ticket is about adding authentication to them.

┆Issue is synchronized with this [Jira Task](https://k8ssandra.atlassian.net/browse/K8SSAND-1135) by [Unito](https://www.unito.io)
┆epic: Reaper Controller Integration
┆fixVersions: k8ssandra-operator-v1.0.0
┆friendlyId: K8SSAND-1135
┆priority: Medium

### PR DESCRIPTION
**Pull request by adejanovski**
**What this PR does**:
Enables auth for the Reaper web UI.
This PR will create a new secret for the Reaper Web UI username/password, unless a secret ref is set in the CR.
This will add env variables to the Reaper container with the username, password, and enable auth boolean.

**Which issue(s) this PR fixes**:
Fixes #276

**Checklist**
- [x] Changes manually tested
- [x] Automated Tests added/updated
- [ ] Documentation added/updated
- [ ] CHANGELOG.md updated (not required for documentation PRs)
- [ ] CLA Signed:  [DataStax CLA](https://cla.datastax.com/)

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -161,11 +161,11 @@ single-up: cleanup build manifests kustomize docker-build create-kind-cluster ki
 single-reload: build manifests kustomize docker-build kind-load-image cert-manager
 	kubectl config use-context kind-k8ssandra-0
 	$(KUSTOMIZE) build config/deployments/control-plane$(DEPLOY_TARGET) | kubectl apply --server-side --force-conflicts -f -
-	kubectl delete pod -l control-plane=k8ssandra-operator -n k8ssandra-operator
-	kubectl rollout status deployment k8ssandra-operator -n k8ssandra-operator
+	kubectl delete pod -l control-plane=k8ssandra-operator -n $(NS)
```
Comment (adutra) [inline Makefile:164]: I'm not sure this change was intentional?

File: Makefile [hunk 2]
```diff
@@ -354,3 +354,8 @@ install-kuttl:
   	ARCH="$$(uname -m | sed -e 's/\(arm\)\(64\)\?.*/\1\2/' -e 's/aarch64$$/arm64/')" ; \
 	curl -LO https://github.com/kudobuilder/kuttl/releases/download/v0.11.1/kuttl_0.11.1_$${OS}_$${ARCH}.tar.gz ; \
 	tar -zxvf kuttl_0.11.1_$${OS}_$${ARCH}.tar.gz ; 
+
+# Regenerate the mocks using mockery
+mocks:
```
Comment (adutra) [inline Makefile:359]: Nice!

File: controllers/reaper/reaper_controller.go [hunk 3]
```diff
@@ -271,63 +273,95 @@ func (r *ReaperReconciler) reconcileService(
...
+
+	authVars := append(cqlVars, jmxVars...)
+	authVars = append(authVars, uiVars...)
+	return authVars, nil
 }
 
-func (r *ReaperReconciler) collectJmxAuthVars(ctx context.Context, actualReaper *reaperapi.Reaper, logger logr.Logger) ([]*corev1.EnvVar, error) {
-	if len(actualReaper.Spec.JmxUserSecretRef.Name) > 0 {
-		secretKey := types.NamespacedName{Namespace: actualReaper.Namespace, Name: actualReaper.Spec.JmxUserSecretRef.Name}
+func (r *ReaperReconciler) collectAuthVarsForType(ctx context.Context, actualReaper *reaperapi.Reaper, logger logr.Logger, authType string) ([]*corev1.EnvVar, error) {
```
Comment (adutra) [inline controllers/reaper/reaper_controller.go:336]: Nice refactoring.

File: controllers/reaper/reaper_controller.go [hunk 4]
```diff
@@ -271,63 +273,95 @@ func (r *ReaperReconciler) reconcileService(
...
+			if err = manager.AddClusterToReaper(ctx, actualDc); err != nil {
+				logger.Error(err, "failed to register cluster with reaper")
+				return ctrl.Result{RequeueAfter: r.DefaultDelay}, err
+			}
 		}
 	}
 	return ctrl.Result{}, nil
 }
 
+func (r *ReaperReconciler) getReaperUISecret(ctx context.Context, actualReaper *reaperapi.Reaper, logger logr.Logger) (string, string, error) {
```
Comment (adutra) [inline controllers/reaper/reaper_controller.go:297]: Nit: `getReaperUICredentials` would be a more accurate name.

File: controllers/reaper/reaper_controller_test.go [hunk 5]
```diff
@@ -378,20 +392,75 @@ func testCreateReaperWithAuthEnabled(t *testing.T, ctx context.Context, k8sClien
...
-	assert.Equal(t, "password", envVars[len(envVars)-4].ValueFrom.SecretKeyRef.Key)
-	assert.Equal(t, "REAPER_CASS_AUTH_ENABLED", envVars[len(envVars)-3].Name)
-	assert.Equal(t, "true", envVars[len(envVars)-3].Value)
-	assert.Equal(t, "REAPER_JMX_AUTH_USERNAME", envVars[len(envVars)-2].Name)
-	assert.Equal(t, "top-secret-jmx", envVars[len(envVars)-2].ValueFrom.SecretKeyRef.LocalObjectReference.Name)
-	assert.Equal(t, "username", envVars[len(envVars)-2].ValueFrom.SecretKeyRef.Key)
-	assert.Equal(t, "REAPER_JMX_AUTH_PASSWORD", envVars[len(envVars)-1].Name)
-	assert.Equal(t, "top-secret-jmx", envVars[len(envVars)-1].ValueFrom.SecretKeyRef.LocalObjectReference.Name)
-	assert.Equal(t, "password", envVars[len(envVars)-1].ValueFrom.SecretKeyRef.Key)
+	assert.True(t, envVarExists(envVars, "REAPER_CASS_AUTH_USERNAME"), "Cassandra auth username env var not found")
```
Comment (adutra) [inline controllers/reaper/reaper_controller_test.go:395]: Much better 👍

File: pkg/reaper/secrets.go [hunk 6]
```diff
@@ -21,6 +25,16 @@ var EnableCassAuthVar = &corev1.EnvVar{
 	Value: "true",
 }
 
+var EnableAuthVar = &corev1.EnvVar{
```
Comment (adutra) [inline pkg/reaper/secrets.go:28]: I think this is unused now?

File: Makefile (hunk 1 above)
Comment (adejanovski) [reply Makefile:164]: it was actually. The `$(NS)` version will cover cases where the deployment is done on another namespace than `k8ssandra-operator` which is the default one.

File: Makefile (hunk 2 above)
Comment (adejanovski) [reply Makefile:359]: thanks

File: controllers/reaper/reaper_controller.go (hunk 3 above)
Comment (adejanovski) [reply controllers/reaper/reaper_controller.go:336]: 🤗

File: controllers/reaper/reaper_controller.go (hunk 4 above)
Comment (adejanovski) [reply controllers/reaper/reaper_controller.go:297]: sounds good, I'll change it.

File: controllers/reaper/reaper_controller_test.go (hunk 5 above)
Comment (adejanovski) [reply controllers/reaper/reaper_controller_test.go:395]: 👍

File: pkg/reaper/secrets.go (hunk 6 above)
Comment (adejanovski) [reply pkg/reaper/secrets.go:28]: It is used [here](https://github.com/adejanovski/k8ssandra-operator/blob/enable-reaper-ui-auth/controllers/reaper/reaper_controller.go#L348).

File: Makefile (hunk 1 above)
Comment (adutra) [reply Makefile:164]: Yeah, after our recents offline conversations about the Makefile I understood the usage of that variable, and indeed it seems to be missing from this specific target.

File: pkg/reaper/secrets.go (hunk 6 above)
Comment (adutra) [reply pkg/reaper/secrets.go:28]: Right, sorry.

File: Makefile (hunk 1 above)
Comment (adutra) [reply Makefile:164]: Unfortunately with your changes and a non-default namespace, the target fails:

```
make NS=custom single-reload
[...]
kubectl -n custom delete pod -l control-plane=k8ssandra-operator
No resources found
kubectl -n custom rollout status deployment k8ssandra-operator
Error from server (NotFound): namespaces "custom" not found
make: *** [single-reload] Error 1
```

**Comment by adejanovski** [comment]
@adutra, the PR is ready for another review.
I was able to fix more issues in the `MultiDcAuthOnOff/TestReaper` e2e tests, but there are still some failures on Stargate which don't seem to be related to the changes in this PR.
Let me know what you think.

File: test/e2e/reaper_test.go [hunk 7]
```diff
@@ -253,10 +261,15 @@ func testRemoveReaperFromK8ssandraCluster(
 	checkReaperK8cStatusReady(t, f, ctx, kcKey, dcKey)
 }
 
-func testReaperApi(t *testing.T, ctx context.Context, k8sContextIdx int, clusterName, keyspace string) {
+func testReaperApi(t *testing.T, ctx context.Context, k8sContextIdx int, clusterName, keyspace string, f *framework.E2eFramework, namespace string, authEnabled bool) {
```
Comment (adutra) [inline test/e2e/reaper_test.go:264]: Minor: could you keep the same order of parameters used in other functions? that is,

```
t *testing.T, 
f *framework.E2eFramework, 
ctx context.Context, 
k8sContextIdx int,
namespace string, 
clusterName string,
<others>
```

File: test/e2e/reaper_test.go [hunk 8]
```diff
@@ -122,18 +130,18 @@ func createMultiReaper(t *testing.T, ctx context.Context, namespace string, f *f
...
-
+	clusterName := "test"
 	dcKey := framework.ClusterKey{K8sContext: "kind-k8ssandra-0", NamespacedName: types.NamespacedName{Namespace: namespace, Name: "dc1"}}
 	reaperKey := framework.ClusterKey{K8sContext: "kind-k8ssandra-0", NamespacedName: types.NamespacedName{Namespace: namespace, Name: "reaper1"}}
 
 	checkDatacenterReady(t, ctx, dcKey, f)
 
 	t.Log("create Reaper keyspace")
-	_, err := f.ExecuteCql(ctx, "kind-k8ssandra-0", namespace, "test", "test-dc1-rack1-sts-0",
+	_, err := f.ExecuteCql(ctx, "kind-k8ssandra-0", namespace, clusterName, fmt.Sprintf("%s-dc1-rack1-sts-0", clusterName),
```
Comment (adutra) [inline test/e2e/reaper_test.go:140]: Why this change? The code looks less obvious to me now compared to the previous version.

File: test/e2e/reaper_test.go [hunk 9]
```diff
@@ -40,20 +42,24 @@ func createSingleReaper(t *testing.T, ctx context.Context, namespace string, f *
 
 	t.Run("TestReaperApi[0]", func(t *testing.T) {
 		t.Log("test Reaper API in context kind-k8ssandra-0")
-		testReaperApi(t, ctx, 0, "test", "reaper_db")
+		authEnabled := authEnabled(t, f, ctx, framework.ClusterKey{K8sContext: "kind-k8ssandra-0", NamespacedName: kcKey})
```
Comment (adutra) [inline test/e2e/reaper_test.go:45]: Tests in this file all have auth enabled so this can be simplified to `true`.

File: test/e2e/reaper_test.go [hunk 10]
```diff
@@ -104,11 +110,13 @@ func createMultiReaper(t *testing.T, ctx context.Context, namespace string, f *f
 	defer f.UndeployAllIngresses(t, "kind-k8ssandra-0", namespace)
 	defer f.UndeployAllIngresses(t, "kind-k8ssandra-1", namespace)
 
+	authEnabled := authEnabled(t, f, ctx, framework.ClusterKey{K8sContext: "kind-k8ssandra-0", NamespacedName: kcKey})
```
Comment (adutra) [inline test/e2e/reaper_test.go:113]: Same here, it will always be true. In fact I'm not sure we need the `authEnabled` function at all.

File: test/testdata/fixtures/multi-dc-reaper/k8ssandra.yaml [hunk 11]
```diff
@@ -16,9 +26,11 @@ spec:
   reaper:
     keyspace: reaper_ks # custom name
     cassandraUserSecretRef:
-      name: reaper-cql-secret # pre-existing secret
+      name: test-cql-secret # pre-existing secret
```
Comment (adutra) [inline test/testdata/fixtures/multi-dc-reaper/k8ssandra.yaml:29]: Why this change? The name can be anything.

File: test/testdata/fixtures/reaper/cassdc.yaml [hunk 12]
```diff
@@ -61,12 +61,12 @@ spec:
             - name: "REAPER_JMX_USERNAME"
               valueFrom:
                 secretKeyRef:
-                  name: reaper-jmx-secret
+                  name: test-jmx-secret
```
Comment (adutra) [inline test/testdata/fixtures/reaper/cassdc.yaml:64]: Why these changes?

File: test/testdata/fixtures/reaper/reaper-cql-secret.yaml [hunk 13]
```diff
@@ -1,7 +1,7 @@
 apiVersion: v1
 kind: Secret
 metadata:
-  name: reaper-cql-secret
+  name: test-cql-secret
```
Comment (adutra) [inline test/testdata/fixtures/reaper/reaper-cql-secret.yaml:4]: Again, not sure changing the name is going to be useful.

File: test/testdata/fixtures/reaper/reaper-jmx-secret.yaml [hunk 14]
```diff
@@ -1,7 +1,7 @@
 apiVersion: v1
 kind: Secret
 metadata:
-  name: reaper-jmx-secret
+  name: test-jmx-secret
```
Comment (adutra) [inline test/testdata/fixtures/reaper/reaper-jmx-secret.yaml:4]: Same...

File: test/testdata/fixtures/reaper/reaper.yaml [hunk 15]
```diff
@@ -7,9 +7,11 @@ spec:
     # TODO check multi-namespace when cluster-scope is ready
     name: dc1
   cassandraUserSecretRef:
-    name: reaper-cql-secret
+    name: test-cql-secret
```
Comment (adutra) [inline test/testdata/fixtures/reaper/reaper.yaml:10]: Same.

**Review by adutra — CHANGES_REQUESTED** [review]

File: test/testdata/fixtures/multi-dc-reaper/k8ssandra.yaml (hunk 11 above)
Comment (adutra) [reply test/testdata/fixtures/multi-dc-reaper/k8ssandra.yaml:29]: Oh I see, you hardcoded the secret name in `getReaperUiSecret`. I think you should remove that function completely. The test code knows what the secret name is, so you just create a `ClusterKey` object with the right name and call `getSecret`.

**Review by adutra — CHANGES_REQUESTED** [review]

File: apis/reaper/v1alpha1/reaper_types.go [hunk 16]
```diff
@@ -172,6 +172,10 @@ type ReaperClusterTemplate struct {
 	// and "password".
 	// +optional
 	JmxUserSecretRef corev1.LocalObjectReference `json:"jmxUserSecretRef,omitempty"`
+
+	// Defines the secret which contains the shiro.ini file controlling the Reaper UI and REST API authentication.
+	// +optional
+	ReaperUiSecretRef corev1.LocalObjectReference `json:"reaperUiSecretRef,omitempty"`
```
Comment (adutra) [inline apis/reaper/v1alpha1/reaper_types.go:178]: For consistency I suggest renaming this field to `UiUserSecretRef`

File: Makefile [hunk 17]
```diff
@@ -164,8 +164,8 @@ single-reload: build manifests kustomize docker-build kind-load-image cert-manag
 	kubectl delete pod -l control-plane=k8ssandra-operator -n k8ssandra-operator
 	kubectl rollout status deployment k8ssandra-operator -n k8ssandra-operator
 ifeq ($(DEPLOYMENT), cass-operator-dev)
-	kubectl -n $(NS) delete pod -l name=cass-operator
-	kubectl -n $(NS) rollout status deployment cass-operator-controller-manager
+	kubectl -n k8ssandra-operator delete pod -l name=cass-operator
```
[the remark this answers is not in the document]
Comment (adejanovski) [reply Makefile:167]: you're right, it shouldn't be part of that PR.

File: apis/reaper/v1alpha1/reaper_types.go (hunk 16 above)
Comment (adejanovski) [reply apis/reaper/v1alpha1/reaper_types.go:178]: 👍

File: controllers/reaper/reaper_controller.go [hunk 18]
```diff
@@ -271,63 +273,95 @@ func (r *ReaperReconciler) reconcileService(
...
+			}
 		}
 	}
 	return ctrl.Result{}, nil
 }
 
+func (r *ReaperReconciler) getReaperUICredentials(ctx context.Context, actualReaper *reaperapi.Reaper, logger logr.Logger) (string, string, error) {
+	secretKey := types.NamespacedName{Namespace: actualReaper.Namespace, Name: actualReaper.Spec.ReaperUiSecretRef.Name}
+	if secret, err := r.getSecret(ctx, secretKey); err != nil {
+		if errors.IsNotFound(err) {
```
[the remark this answers is not in the document]
Comment (adejanovski) [reply controllers/reaper/reaper_controller.go:300]: It was handling the case where the secret name didn't exist (auth is disabled).
I've added a check based on your previous comment, so I can change this part now and return an error.

File: pkg/reaper/secrets.go [hunk 19]
```diff
@@ -30,21 +44,30 @@ func DefaultJmxUserSecretName(clusterName string) string {
...
+func GetAuthEnvironmentVars(secret *corev1.Secret, authType string) (*corev1.EnvVar, *corev1.EnvVar, error) {
+	switch authType {
+	case "cql":
+		return secretToEnvVars(secret, cassAuthEnvUsernameName, cassAuthEnvPasswordName)
+	case "jmx":
+		return secretToEnvVars(secret, jmxAuthEnvUsernameName, jmxAuthEnvPasswordName)
+	case "ui":
+		return secretToEnvVars(secret, uiAuthEnvUsernameName, uiAuthEnvPasswordName)
+	default:
+		return nil, nil, fmt.Errorf("Unsupported auth type %s", authType)
```
[the remark this answers is not in the document]
Comment (adejanovski) [reply pkg/reaper/secrets.go:60]: 👍

File: apis/reaper/v1alpha1/reaper_types.go [hunk 20]
```diff
@@ -172,6 +172,10 @@ type ReaperClusterTemplate struct {
 	// and "password".
 	// +optional
 	JmxUserSecretRef corev1.LocalObjectReference `json:"jmxUserSecretRef,omitempty"`
+
+	// Defines the secret which contains the shiro.ini file controlling the Reaper UI and REST API authentication.
```
Comment (adutra) [inline apis/reaper/v1alpha1/reaper_types.go:176]: Nit: The comment states that the secret should contain the whole shiro.ini file, however it seems that in our tests, we expect the secret to contain just a `username` and a `password` entry, like other secrets in this project.

**Review by adutra — APPROVED** [review]
\o/
